docs/designers-developers/developers/tutorials/javascript/scope-your-code.md

```diff
@@ -0,0 +1,117 @@
+
+Taking advantage of these two JavaScript features, `first.js` could be scoped as:
+
+```js
+function() {
+	var pluginName = 'MyPlugin';
+	console.log( 'Plugin name is ', pluginName );
+}
+```
+
+`second.js` as:
+
+```js
+function() {
+	var pluginName = 'DifferentPlugin';
+	console.log( 'Plugin name is ', pluginName );
+}
+```
+
+And `third.js`:
+
+```js
+function() {
+	console.log( 'Plugin name is ', pluginName );
+}
+```
+
+With this trick, the different files won't override each other's variables. Unfortunately, they also won't work as expected, because these functions are being called by no one. We've only _defined_ the functions; we haven't _executed_ them yet.
+
+## Automatically execute anonymous functions
+
+It turns out there are a few ways to execute anonymous functions in JavaScript, but the most popular is this:
+
+```js
+( function() {
+	// your code goes here
+} )( )
+```
+
+You wrap your function between parentheses, and then call it like any other named function. This pattern is known as [Immediately-Invoked Function Expression](http://benalman.com/news/2010/11/immediately-invoked-function-expression/), or IIFE for short.
+
+This is `first.js` written as an IIFE:
+
+```js
+( function() {
+	var pluginName = 'MyPlugin';
+	console.log( 'Plugin name is ', pluginName );
+} )( )
+```
+
+And this is `second.js`:
+
+```js
+( function() {
+	var pluginName = 'DifferentPlugin';
+	console.log( 'Plugin name is ', pluginName );
+} )( )
+```
+
+And this is `third.js`:
+
+```js
+( function() {
+	console.log( 'Plugin name is ', pluginName );
+} )( )
+```
+
+The code in `first.js` and `second.js` is unaffected by other variables in the global scope, so it's safe and deterministic.
+
+On the other hand, `third.js` doesn't declare a `pluginName` variable, but needs to be provided one. IIFEs still allow you to take a variable from the global scope and pass it into your function. Provided that there was a global `window.pluginName` variable, we could rewrite `third.js` as:
+
+```js
+( function( name ) {
+	console.log( 'Plugin name is ', name );
+} )( window.pluginName )
+```
+
+## Future changes
+
+At the beginning we mentioned that:
+
+> Historically, JavaScript files loaded in a web page share the same scope.
+
+Notice the _historically_.
+
+JavaScript has evolved quite a bit since its creation. As of 2015, the language supports modules, also known as _ES6 modules_, that introduce separate scope per file: a global variable in `first.js` wouldn't be exposed to `second.js`. This feature is already [supported by modern browsers](https://caniuse.com/#feat=es6-module), but not all of them do. If your code needs to run in browsers that don't support modules, your last resort is using IIFEs.
```

Comment:
My intent was making this as free as jargon as possible as to make it more palatable to people not exposed to modern JavaScript just yet, that's why I only hinted at ES6 by the end of the file.

One way I think this tutorial could be expanded is by adding a section that includes a minimum viable mechanism to compile from ES6 to ES5 (including JSX). I was tempted to adapt [my own dumb webpack](http://nosolosoftware.com/on-javascript-modules/) explanation for the handbook. Upon sleeping on it, I think that a better use of everyone's time (the writer and the reader) will be to walk the reader through a minimal setup instead - that way we give developers an easy and practical way to get started and avoid the analysis paralysis.

Reply:
A separate tutorial for build setup would be good, but we should probably wait for @wordpress/scripts improvements first and use it instead of using the tools directly cc @gziolo. I was more thinking of a random person arriving into this tutorial. They might already use ES6. So just a small note at the beginning could be useful IMO.

Reply:
Yes, I think we will need to add a tutorial about using `wp-scripts` in your project. There are three things that need happen before we can start promoting it:
- [Scripts: Add support for start and build scripts to the package](https://github.com/WordPress/gutenberg/pull/12837)
- [Re-use webpack and babel config for build script](https://github.com/WordPress/gutenberg/pull/12685)
- All packages needs to be published to npm

Reply:
I'd love to! Ping me in any PR that lacks a reviewer and I'll be there. When pieces are in place, I can also pair to write the tutorial.